### PR TITLE
Sam Warren Candidate Assignment

### PR DESCRIPTION
**Pull request by STSWarren**
Candidate assignment finished by Sam Warren

### CODE REVIEW

File: CsvFindings.txt [hunk 1]
```diff
@@ -0,0 +1,11 @@
+Multiple Postal communities are linked to multiple political communities and vice versa. 
+
+The zip and zip addition are the compound key for a postal community.
+
+The short name and long name of the Political communities are the exact same in the GDE csv.
+
+Many different postal communities share the same name, including Lausanne, who is, surprisingly, in the political community Lausanne. But only 10% of this Lausanne is in Lausanne.
+
+Unless “%_IN_GDE” means something else, some political communities must slightly overlap. In the case of Bercher, it is 100% in Bercher but also 0.76% in Boulens and 0.21% Montanaire
```
Comment (robertblust) [inline CsvFindings.txt:9]: the percentage means how much of a postalcommunity (zip 1 zipAddition) is part of a politicalCommunity

`curl -X GET "https://platform-prod.3ap.ch/api/zipcode-service/postal-communities/search?param=Bercher" -H  "accept: */*"`

thus 

```
   "zipCode": "1038",
    "zipCodeAddition": "0"
```

is 100% in

```
"politicalCommunity": {
      "number": "5512",
      "name": "Bercher",
      "shortName": "Bercher",
      "lastUpdate": "2006-09-01"
    }
```

File: src/main/java/ch/aaap/assignment/Application.java [hunk 2]
```diff
@@ -36,46 +45,135 @@ public Model getModel() {
...
+    LocalDate lastUpdateOfPoliticalCommunity = LocalDate.now();
+    List<PostalCommunity> postalCommunityMatchList = 
+      model.getPostalCommunities().stream().filter( postalCommunity -> 
+        postalCommunity.getName().equals( postalCommunityName )).collect( Collectors.toList() );
+
+    if( postalCommunityMatchList.isEmpty() ){
+      throw new IllegalArgumentException("No postal community with that name exists.");
+    }
+
+    if( postalCommunityMatchList.size() > 1){
```
Comment (robertblust) [inline src/main/java/ch/aaap/assignment/Application.java:149]: this is related to the question you send by email. e.g when you look for name Lausanne. yes you can not assigned it to a unique postalCommunity 🤔 good catch 👍

**Review by robertblust — COMMENTED** [review]
@STSWarren thanks for submitting the candidate assignment. All test cases pass.

See some my comments. 
And overall the code formatting is not according google styleguide 😞 

You will be informed soon by Simona

File: CsvFindings.txt [hunk 3]
```diff
@@ -0,0 +1,11 @@
+Multiple Postal communities are linked to multiple political communities and vice versa.
```
[the remark this answers is not in the document]
Comment (STSWarren) [reply CsvFindings.txt:1]: Hi Robert, is this in relation to how %_in_GDE works? To give an example, the postal community "Prilly" with the Zip 1008 and zip addition "0" has a line in the csv with GDENR 5589 and another line with GDENR 5591. Prilly is 100% in GDENR 5589 and 0.06% in GDENR 5591. However, the CSVUtil does not discard any lines based on their percentages and also doesnt store the percentages either. In the end, you are left with 1008,0 linked to 5589 and 5591. 
example 1

This is also true of zip code and addition 1015,0 which is linked to GDENR 5627, 5635 and 5648 at the percentages 1.21, 3.58, 0.14 respectively. 
example 2

In reverse, we can see with zip and additions (1005,0), (1006,0), (1007,0) are all, in some part, in GDENR 5586, so political communities are connected to multiple postal communities too.
example 3

File: src/main/java/ch/aaap/assignment/Application.java (hunk 2 above)
Comment (STSWarren) [reply src/main/java/ch/aaap/assignment/Application.java:149]: I wanted to just quickly say thank you for reviewing the code so quickly, as well as your feedback. I'll have to review the Google styleguide again to try figure out what I missed! 

I hope to hear from you soon and thanks again.

File: CsvFindings.txt (hunk 3 above)
[the remark this answers is not in the document]
Comment (robertblust) [reply CsvFindings.txt:1]: i see it that the  %_IN_GDE is like 100% of the GDE 5589 are covered by the postalcommunity 1008,0

[the remark this answers is not in the document]
Comment (STSWarren) [reply CsvFindings.txt:1]: I see, but then CSVUtil does not distinguish between the two, as it reads all the data into the set of CSVPostalCommunity objects without this percentage. As a result a CSVPostalCommunity object will exist with
zip 1008, zipAddition 0, name Prilly, cantonCode VD, politicalCommunityNumber 5589 and politicalCommunityShortName Prilly 
and another with 
zip 1008, zipAddition 0, name Prilly, cantonCode VD, politicalCommunityNumber 5591 and politicalCommunityShortName Renens (VD). 

If we create a data model where only one postalCommunity can be linked to one politicalCommunity, we must choose if 5589 is linked to 1008,0 or if 5591 is. Without the %_in_GDE variable, we either have to do this at random or based on whichever of the objects comes first when mapping to PostalCommunity type objects in the list.

**Comment by robertblust** [comment]
@STSWarren the goal is not that you fix all the things I commented. The comment are more to appreciate the effort you made.

**Pull request closed without merging**